# cutfrequency
<p align="center">
  <img src="https://user-images.githubusercontent.com/28807444/130489546-74e51a39-23d4-494d-b810-29eb42efa2c1.jpg" align='left' height='200' width='250' />
</p>
Simple script to calculate cut frequency around transcription factor binding sites using BAM files (pair-end reads): 
greenCUT&RUN or CUT&RUN protocol



Citation: 
	
	Nizamuddin et al (2021), Integrating quantitative proteomics 
	with accurate genome profiling of transcription factors by 
	greenCUT&RUN. Nucleic Acids Research, 46(9) e49 
	(https://doi.org/10.1093/nar/gkab038)

REQUIREMENTS:

	Samtools
	Bedtools
	R, optparse, data.table. svMisc
	perl, List::Util
	Linux/Unix operating system


USAGE:

	Rscript run.R [options]

Options:
	
	-a CHARACTER, --bam1=CHARACTER
		name of bam file: experiment (without .bam extension)

	-b CHARACTER, --bam2=CHARACTER
		name of bam file: control (without .bam extension)

	-c CHARACTER, --bam3=CHARACTER
		name of bam file - spikeIn: experiment (without .bam extension)

	-d CHARACTER, --bam4=CHARACTER
		name of bam file - spikeIn: control (without .bam extension)

	-m CHARACTER, --motifFile=CHARACTER
		File having location of motif; generated by Homer

	-x NUMERIC, --centreOfmotif=NUMERIC
		Desired centre of motif on positive DNA strand

	-s CHARACTER, --script_folder=CHARACTER
		Folder location where filepreparation.pl of the current tool is present

	-n NUMERIC, --thread=NUMERIC
		Number of thread

	-h, --help
		Show this help message and exit
		
	-p CHARACTER, --peakFile=CHARACTER
		If the number of the motifs are higher in the genome and do not have enough memory, 
		then might be interested in exploring cut-frequency at only peaks. 
		If yes then, provide peakFile in the bed format **OPTIONAL**

EXAMPLE:

	A) To generate motif file of the protein of interest proceed as follows (install homer):
	
		1) download file contains PWM of the motifs
			http://homer.ucsd.edu/homer/custom.motifs
		2) find protein of interest in the custom.motifs and 
		copy-paste the PWM in new file (e.g. tp53.pwm or fos.pwm)
		
		3) run following command to generate bed file containing location 
		of whole genome motifs (http://homer.ucsd.edu/homer/motif/genomeWideMotifScan.html)
		
		scanMotifGenomeWide.pl fos.pwm hg38 | cut -f2- > FOS-motifs.txt
		or
		scanMotifGenomeWide.pl fos.pwm genome.fa | cut -f2- > FOS-motifs.txt
	
	B) Rscript ./run.R -a ./Example/experiment -b ./Example/control -c ./Example/SpikeIn-experiment
	-d ./Example/SpikeIn-control  -m ./Example/FOS-motifs.txt -x 7 --script_folder ./ -n 20 

OUTPUT:

	experiment.bed3 - Genomic coordinates of the pair-end reads
	
	experiment.input - Experiment's read overlapping with motifs
	
	control.bed3 - Genomic coordinates of the pair-end reads
	
	control.input - Control's read overlapping with motifs
	
	SpikeIn-experiment.bed3 - Genomic coordinates of the pair-end reads. 
				For counting Spike-In reads in the experiment.
				
	SpikeIn-control.bed3 - Genomic coordinates of the pair-end reads. 
				For counting Spike-In reads in the control.
				
	control.counts.txt: Count of read ends for each motif's location. 
				-100 to 100 bps from centre of motif 
				
	experiment.counts.txt:  Count of read ends for each motif's location. 
				-100 to 100 bps from centre of motif 
				
	experiment.normalize.txt: Spike-in normalized read ends in experiment 
				compare to control, for each motif's location.
				-100 to 100 bps from centre of motif 
				
	experiment.cutfrequency.txt: Cutfrquency at each location from 
				-100 to 100
				
	experiment.jpeg: visualization

