if(!require(optparse)){
        install.packages("optparse")
    library(optparse)
}
if(!require(data.table)){
        install.packages("data.table")
    library(data.table)
}
if(!require(svMisc)){
        install.packages("svMisc")
    library(svMisc)
}

option_list = list(
make_option(c("-a", "--bam1"), type="character", default=NA,help="name of bam file: experiment (without .bam extension)", metavar="character"),
make_option(c("-b", "--bam2"), type="character", default=NA,help="name of bam file: control (without .bam extension)", metavar="character"),
make_option(c("-c", "--bam3"), type="character", default=NA,help="name of bam file - spikeIn: experiment (without .bam extension)", metavar="character"),
make_option(c("-d", "--bam4"), type="character", default=NA,help="name of bam file - spikeIn: control (without .bam extension)", metavar="character"),
make_option(c("-m", "--motifFile"), type="character", default=NA,help="File having location of motif; generated by Homer", metavar="character"),
make_option(c("-x", "--centreOfmotif"), type="integer", help="Desired centre of motif on positive DNA strand", metavar="numeric"),
make_option(c("-s", "--script_folder"), type="character", default=NA,help="Folder location with other assessory scripts of this tool", metavar="character"),
make_option(c("-n", "--thread"), type="integer", default=2,help="Number of thread",metavar="numeric")

);

parser<- OptionParser(option_list=option_list)
opt <- parse_args(parser)

bam1=opt$bam1
bam2=opt$bam2
bam3=opt$bam3
bam4=opt$bam4
mymotif=opt$motifFile
mythread=opt$thread
myfolder=opt$script_folder
centre=opt$centreOfmotif
	
print ('=============================')
print ('=== Cutfrequency calculation ')
print ('=== snizam001@gmail.com   ===')
print ('=============================')

if(!(file.exists(paste(myfolder,"/filePreparation.pl",sep="")))) {
	print(paste("filePreparation.pl file is not present in ",myfolder,sep=""));
	stop()
}

#---- data preparation: part1
print('==== file preparation: bamfile 1')
f=paste("perl ",myfolder,"/filePreparation.pl ", bam1," ", mymotif , " ", mythread,sep="" )
x = system(f)

if (x != 0)
	{ 
		stop()
	}

print('==== file preparation: bamfile 2')
f=paste("perl ",myfolder,"/filePreparation.pl ", bam2," ", mymotif, " " , mythread,sep="" )
x = system(f)

if (x != 0)
	{ 
		stop()
	}

print('==== file preparation: bamfile 3')
f=paste("perl ",myfolder,"/filePreparation.pl ", bam3," ", mymotif, " " , mythread,sep="" )
x = system(f)
if (x != 0)
	{ 
		stop()
	}

system(paste("rm ", bam3, ".input",sep=""))
print('==== file preparation: bamfile 4')
f=paste("perl ",myfolder,"/filePreparation.pl ", bam4," ", mymotif, " " , mythread,sep="" )
x = system(f)
if (x != 0)
	{ 
		stop()
	}
system(paste("rm ", bam4, ".input",sep=""))

#---- data preparation: part2
d <- fread(paste(bam1,'.input',sep=''))
d <- data.frame(d)
pos_d <- d[d[,4]=="+",]
pos_d$end5 = pos_d[,8] - (pos_d[,2]+(centre-1))
pos_d$end3 = pos_d[,9] - (pos_d[,2]+(centre-1))
neg_d <- d[d[,4]=="-",]
neg_d$end5 = neg_d[,9] - (neg_d[,2]+(centre-1))
neg_d$end3 = neg_d[,8] - (neg_d[,2]+(centre-1))
rbind(pos_d,neg_d) -> values
values[order(values[,1],values[,2]),] -> values
experiment = values
#--
d <- fread(paste(bam2,'.input',sep=''))
d <- data.frame(d)
pos_d <- d[d[,4]=="+",]
pos_d$end5 = pos_d[,8] - (pos_d[,2]+(centre-1))
pos_d$end3 = pos_d[,9] - (pos_d[,2]+(centre-1))
neg_d <- d[d[,4]=="-",]
neg_d$end5 = neg_d[,9] - (neg_d[,2]+(centre-1))
neg_d$end3 = neg_d[,8] - (neg_d[,2]+(centre-1))
rbind(pos_d,neg_d) -> values
values[order(values[,1],values[,2]),] -> values
control = values



#--
d <- fread(paste(bam3,'.bed3',sep=''))
d <- data.frame(d)
nrow(d) -> spike_experiment
#--
d <- fread(paste(bam4,'.bed3',sep=''))
d <- data.frame(d)
nrow(d) -> spike_control
#--
d <- fread(paste(bam1,'.bed3',sep=''))
d <- data.frame(d)
nrow(d) -> experiment_reads
#--
d <- fread(paste(bam2,'.bed3',sep=''))
d <- data.frame(d)
nrow(d) -> control_reads
#--
d <- fread(mymotif)
d <- data.frame(d)
nrow(d) -> motifs_number
#--
spike_ratio=spike_experiment/spike_control
print (paste('=== Spike-In ratio in experiment vs control is ',spike_ratio,sep=''))



#--- Nearest cut
print ('=== Finding nearest cut for each reads ')
print ('=== Experiment ')
as.numeric(ifelse ((experiment$end5^2)^(1/2) < (experiment$end3^2)^(1/2) , experiment$end5, experiment$end3)) -> experiment[,15]
print ('=== Control ')
as.numeric(ifelse ((control$end5^2)^(1/2) < (control$end3^2)^(1/2) , control$end5, control$end3)) -> control[,15]
experiment$region=paste(experiment[,1],':',experiment[,2],'-',experiment[,3],sep='');
control$region=paste(control[,1],':',control[,2],'-',control[,3],sep='');



#-- counting
print ('=== Counting cuts per region ')
table(experiment$region,experiment[,15]) -> mytable_expr
mytable_expr[,as.character(seq(-100,100,1))]->mytable_expr
#--
table(control$region,control[,15]) -> mytable_ctrl
mytable_ctrl[,as.character(seq(-100,100,1))]->mytable_ctrl


print ('=== output of counts in table ')
write.table(mytable_ctrl[rowSums(mytable_ctrl) > 1,],paste(bam2,'.counts.txt',sep=''),quote=F,sep='\t',col.names=F)
write.table(mytable_expr[rowSums(mytable_expr) > 1,],paste(bam1,'.counts.txt',sep=''),quote=F,sep='\t',col.names=F)
#---- normalization


print ('=== Normalization ')
common = intersect(rownames(mytable_ctrl),rownames(mytable_expr))
specific2expr = setdiff(rownames(mytable_expr),rownames(mytable_ctrl))
mytable_expr=1000000*((mytable_expr/experiment_reads)/motifs_number)
mytable_ctrl = mytable_ctrl*spike_ratio
mytable_ctrl=1000000*((mytable_ctrl/control_reads)/motifs_number)
print ('=== Comparing with control ')
out = mytable_expr[common,] - mytable_ctrl[common,]
out = rbind(out,mytable_expr[specific2expr,])
write.table(out[rowSums(out) > 0,],paste(bam1,'.normalize.txt',sep=''),quote=F,sep='\t',col.names=F)
z<-colSums(out)
out_frequency=data.frame(Distance=as.numeric(names(z)),Frequency=as.numeric(z))
write.table(out_frequency,paste(bam1,'.cutfrequency.txt',sep=''),quote=F,sep='\t',col.names=F)
print(z)
maximum=max(z)
minimum=min(z)
jpeg(paste(bam1,'.jpeg',sep=''),unit='in',res=300,height=4.5,width=4.5)
plot(as.numeric(names(z)),z,ylim=c(minimum,maximum),type='n',xlab='Position (bps)',ylab='Cut frequency')
cc = nchar(experiment[1,6])
right_cc = cc - centre
left_cc = -1 * (cc - (right_cc + 1))
rect(left_cc,minimum-500,right_cc,maximum+500,col='grey80',border = NA)
for(i in seq(-100,100,10)) {abline(v=i,lty=2,col='grey72')}
lines(as.numeric(names(z)),z,col='black',type='o')
#lines(as.numeric(names(colSums(mytable_ctrl))),colSums(mytable_ctrl),type='l',col='orange')
dev.off()
print ('=== Finished ')
print ('=============================')



